File: typescript/src/ships/gunsights.ts
/**
 * Fixed-weapon frontal gunsights for every player-flyable ship.
 *
 * The catalogue keeps only the geometry needed to place each hardpoint at any target
 * range: a horizontal and vertical offset from the cockpit, both in metres. Use
 * {@link projectGunsight} to turn those physical offsets into dimensionless angular
 * tangents suitable for a renderer.
 *
 * The offsets are observed in-game; see
 * [`data/ships/SOURCES.md`](https://github.com/DarkSession/Elite-Dangerous-Almanac/blob/main/data/ships/SOURCES.md).
 *
 * @packageDocumentation
 */

import gunsightsData from '../../../data/ships/gunsights.jsonc' with { type: 'json' };
import { deepFreeze } from '../internal/deep-freeze.js';
import { findByRawKey } from '../internal/registry-index.js';

/**
 * One fixed weapon's camera-relative origin in metres.
 *
 * @remarks
 * The first value is horizontal: positive points right. The second is vertical:
 * positive points up. Divide each value by a positive target range in metres to obtain
 * the corresponding angular tangent; {@link projectGunsight} performs that operation.
 */
export type GunsightOffset = readonly [
    horizontalOffsetMetres: number,
    verticalOffsetMetres: number,
];

/**
 * A hull's fixed-weapon offsets, in exactly the same order as {@link Ship.hardpoints}.
 *
 * @remarks
 * The shared ordering avoids repeating slot names in this compact catalogue. Pair an
 * offset with `ship.hardpoints[index]`, or with the hardpoint entries from
 * `enumerateSlots(ship)` in their returned order, when a renderer needs the mount's
 * size or journal-compatible slot key. **The number in a journal slot key is not the
 * array index**: some hulls skip or reorder those numbers, so resolve a journal key
 * through `enumerateSlots(ship)` before using its index here.
 */
export type ShipGunsight = readonly GunsightOffset[];

/** A projected fixed-weapon point relative to the cockpit's forward direction. */
export interface GunsightPoint {
    /** Dimensionless horizontal angular tangent; positive points right. */
    readonly horizontalTangent: number;
    /** Dimensionless vertical angular tangent; positive points up. */
    readonly verticalTangent: number;
}

/** A hull-symbol-keyed map of gunsights. */
export type ShipGunsightCatalogue = Readonly<Record<string, ShipGunsight>>;

/**
 * Fixed-weapon gunsights for every player-flyable hull, keyed by the hull symbols used
 * by {@link ships!SHIPS | SHIPS}.
 *
 * @remarks
 * The map, each hull array and every offset pair are deeply frozen. It covers every
 * hardpoint of every player-flyable hull. Importing this module adds only the compact
 * offset map; it does not import the ship or outfitting catalogues.
 *
 * @example
 * ```ts
 * import { SHIP_GUNSIGHTS } from '@elite-dangerous-almanac/core/ships/gunsights';
 *
 * SHIP_GUNSIGHTS.SideWinder?.length; // -> 2
 * SHIP_GUNSIGHTS.SideWinder?.[0]; // -> [-2.1956754, -1.166162]
 * ```
 */
export const SHIP_GUNSIGHTS: ShipGunsightCatalogue = deepFreeze(
    gunsightsData as ShipGunsightCatalogue,
);

/**
 * Look up a hull's gunsight by its internal symbol, case-insensitively.
 *
 * @param shipSymbol - Hull symbol, e.g. `"SideWinder"`. Leading/trailing whitespace
 * and case are ignored.
 * @returns The deeply frozen gunsight, or `null` when the hull is unknown.
 * @throws {TypeError} If `shipSymbol` is present and not a string. A nullish value is
 * a miss, like any other unrecognised symbol.
 * @example
 * ```ts
 * import { getShipGunsight } from '@elite-dangerous-almanac/core/ships/gunsights';
 *
 * getShipGunsight(' sidewinder ')?.length; // -> 2
 * getShipGunsight('not_a_ship'); // -> null
 * ```
 */
export function getShipGunsight(shipSymbol: string): ShipGunsight | null {
    return findByRawKey(SHIP_GUNSIGHTS, shipSymbol, 'getShipGunsight: shipSymbol');
}

/**
 * Project a gunsight onto a target plane at a chosen range.
 *
 * @param gunsight - Camera-relative hardpoint offsets in metres, normally returned by
 * {@link getShipGunsight}.
 * @param targetRangeMetres - Finite target-plane distance in metres; must be greater
 * than zero.
 * @returns One dimensionless angular-tangent point per input offset, preserving its
 * order. To place a point on a perspective display, divide `horizontalTangent` and
 * `verticalTangent` by the tangent of the corresponding half field of view; invert the
 * vertical result when screen coordinates increase downwards.
 * @remarks
 * This is fixed, ship-forward geometry. It does not model gimbal or turret tracking,
 * projectile travel, target motion, or head-look.
 * @throws {RangeError} If `targetRangeMetres` is not finite and greater than zero.
 * @example
 * ```ts
 * import {
 *   getShipGunsight,
 *   projectGunsight,
 * } from '@elite-dangerous-almanac/core/ships/gunsights';
 *
 * const gunsight = getShipGunsight('SideWinder')!;
 * projectGunsight(gunsight, 1000)[0];
 * // -> { horizontalTangent: -0.0021956754, verticalTangent: -0.001166162 }
 * ```
 */
export function projectGunsight(
    gunsight: ShipGunsight,
    targetRangeMetres: number,
): readonly GunsightPoint[] {
    if (!Number.isFinite(targetRangeMetres) || targetRangeMetres <= 0) {
        throw new RangeError('projectGunsight: targetRangeMetres must be a finite positive number');
    }
    return gunsight.map(([horizontalOffsetMetres, verticalOffsetMetres]) => ({
        horizontalTangent: horizontalOffsetMetres / targetRangeMetres,
        verticalTangent: verticalOffsetMetres / targetRangeMetres,
    }));
}

File: dotnet/src/EliteDangerousAlmanac/Ships/Gunsights.cs
using System;
using System.Collections.Generic;
using System.Collections.ObjectModel;
using EliteDangerousAlmanac.Internal;

namespace EliteDangerousAlmanac.Ships;

/// <summary>One fixed weapon's camera-relative origin, in metres.</summary>
/// <param name="HorizontalMetres">The horizontal offset. A positive figure points right.</param>
/// <param name="VerticalMetres">The vertical offset. A positive figure points up.</param>
/// <remarks>
/// Divide each figure by a target range in metres to obtain the corresponding angular tangent.
/// </remarks>
public sealed record GunsightOffset(double HorizontalMetres, double VerticalMetres);

/// <summary>A projected fixed-weapon point, relative to the cockpit's forward direction.</summary>
/// <param name="HorizontalTangent">The horizontal angular tangent. A positive figure points right.</param>
/// <param name="VerticalTangent">The vertical angular tangent. A positive figure points up.</param>
/// <remarks>Both figures have no unit.</remarks>
public sealed record GunsightPoint(double HorizontalTangent, double VerticalTangent);

/// <summary>The fixed-weapon frontal gunsights for every player-flyable ship.</summary>
/// <remarks>
/// <para>
/// The catalogue keeps only the geometry needed to place each hardpoint at any target range: a
/// horizontal and a vertical offset from the cockpit, both in metres.
/// </para>
/// <para>
/// A hull's offsets follow its own hardpoints in exactly the same order, which is why this
/// compact catalogue does not repeat the slot names. Pair an offset with the hardpoint at the
/// same position. The number in a journal slot key is not that position: some hulls skip or
/// reorder those numbers, so resolve a journal key through the hull's enumerated slots first.
/// </para>
/// <para>The offsets are observed in-game. See <c>data/ships/SOURCES.md</c>.</para>
/// </remarks>
public static class Gunsights
{
    private static readonly Lazy<IReadOnlyDictionary<string, IReadOnlyList<GunsightOffset>>>
        Catalogue = new(Load);

    /// <summary>Every hull's fixed-weapon offsets, keyed by hull symbol.</summary>
    /// <remarks>
    /// It covers every hardpoint of every player-flyable hull. The map and its lists refuse
    /// mutation.
    /// </remarks>
    public static IReadOnlyDictionary<string, IReadOnlyList<GunsightOffset>> All => Catalogue.Value;

    /// <summary>Looks up a hull's gunsight by its internal symbol, ignoring case.</summary>
    /// <param name="shipSymbol">
    /// The hull symbol. Leading and trailing whitespace and letter case are ignored.
    /// </param>
    /// <returns>The hull's offsets, or <see langword="null"/> where the hull is unknown.</returns>
    public static IReadOnlyList<GunsightOffset>? Find(string? shipSymbol) =>
        RegistryIndex.FindInKeyIndex(All, shipSymbol);

    /// <summary>Projects a gunsight onto a target plane at a chosen range.</summary>
    /// <param name="gunsight">The camera-relative hardpoint offsets, in metres.</param>
    /// <param name="targetRangeMetres">The target-plane distance, in metres, which must be above zero.</param>
    /// <returns>
    /// One angular-tangent point per offset, in the order the offsets came. To place a point on a
    /// perspective display, divide each tangent by the tangent of the corresponding half field of
    /// view, and invert the vertical result where screen coordinates increase downwards.
    /// </returns>
    /// <remarks>
    /// This is fixed, ship-forward geometry. It does not model gimbal or turret tracking,
    /// projectile travel, target motion, or head-look.
    /// </remarks>
    /// <exception cref="ArgumentNullException"><paramref name="gunsight"/> is <see langword="null"/>.</exception>
    /// <exception cref="ArgumentOutOfRangeException">
    /// The target range is not a finite number above zero.
    /// </exception>
    public static IReadOnlyList<GunsightPoint> Project(
        IReadOnlyList<GunsightOffset> gunsight, double targetRangeMetres)
    {
        if (gunsight is null) throw new ArgumentNullException(nameof(gunsight));
        if (double.IsNaN(targetRangeMetres)
            || double.IsInfinity(targetRangeMetres)
            || targetRangeMetres <= 0)
        {
            throw new ArgumentOutOfRangeException(
                nameof(targetRangeMetres),
                targetRangeMetres,
                "The value must be a finite number above 0.");
        }

        var points = new List<GunsightPoint>(gunsight.Count);
        foreach (GunsightOffset offset in gunsight)
        {
            if (offset is null) throw new ArgumentNullException(nameof(gunsight));

            points.Add(new GunsightPoint(
                offset.HorizontalMetres / targetRangeMetres,
                offset.VerticalMetres / targetRangeMetres));
        }

        return new ReadOnlyCollection<GunsightPoint>(points);
    }

    private static ReadOnlyDictionary<string, IReadOnlyList<GunsightOffset>> Load()
    {
        var file = SharedData.Load<Dictionary<string, double[][]>>("data/ships/gunsights.jsonc");
        Dictionary<string, IReadOnlyList<GunsightOffset>> hulls = [];
        foreach (KeyValuePair<string, double[][]> hull in file)
        {
            var offsets = new List<GunsightOffset>(hull.Value.Length);
            foreach (double[] pair in hull.Value)
            {
                offsets.Add(new GunsightOffset(pair[0], pair[1]));
            }

            hulls[hull.Key] = new ReadOnlyCollection<GunsightOffset>(offsets);
        }

        return RegistryIndex.FreezeByRawKey<IReadOnlyList<GunsightOffset>>(hulls);
    }
}
